### PR TITLE
Base64 canonical and URL-friendly encoding/decoding.

### PR DESCRIPTION
**Pull request by mzhurovich**

### CODE REVIEW

File: Bricks/util/base64.h [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+      if (nbits >= 8) {
+        nbits -= 8;
+        result.push_back(char((buf >> nbits) & 0xFF));
+      }
+    }
+    return result;
+  }
+};
+
+}  // namespace base64
```
Comment (dkorolev) [inline Bricks/util/base64.h]: `}  // namespace current::base64` ?

**Comment by dkorolev** [comment]
LGTM!

File: Bricks/util/base64.h [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+namespace current {
+
+struct Base64DecodeException : Exception {
+  using Exception::Exception;
+};
+
+namespace base64 {
+
+constexpr static char encode_map[65] = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
+constexpr static char url_encode_map[65] = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_";
```
Comment (sompylasar) [inline Bricks/util/base64.h]: Please add a comment with a reference to the standard: https://tools.ietf.org/html/rfc3548#section-4

File: Bricks/util/base64.h [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+
+namespace current {
+
+struct Base64DecodeException : Exception {
+  using Exception::Exception;
+};
+
+namespace base64 {
+
+constexpr static char encode_map[65] = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
```
Comment (sompylasar) [inline Bricks/util/base64.h]: Please add a comment with a reference to the standard: https://tools.ietf.org/html/rfc3548#section-3

File: Bricks/util/golden/base64test.base64 [hunk 4]
```diff
@@ -0,0 +1 @@
...aXNoLCBkaXN0cmlidXRlLCBzdWJsaWNlbnNlLCBhbmQvb3Igc2VsbApjb3BpZXMgb2YgdGhlIFNvZnR3YXJlLCBhbmQgdG8gcGVybWl0IHBlcnNvbnMgdG8gd2hvbSB0aGUgU29mdHdhcmUgaXMKZnVybmlzaGVkIHRvIGRvIHNvLCBzdWJqZWN0IHRvIHRoZSBmb2xsb3dpbmcgY29uZGl0aW9uczoKClRoZSBhYm92ZSBjb3B5cmlnaHQgbm90aWNlIGFuZCB0aGlzIHBlcm1pc3Npb24gbm90aWNlIHNoYWxsIGJlIGluY2x1ZGVkIGluIGFsbApjb3BpZXMgb3Igc3Vic3RhbnRpYWwgcG9ydGlvbnMgb2YgdGhlIFNvZnR3YXJlLgoKVEhFIFNPRlRXQVJFIElTIFBST1ZJREVEICJBUyBJUyIsIFdJVEhPVVQgV0FSUkFOVFkgT0YgQU5ZIEtJTkQsIEVYUFJFU1MgT1IKSU1QTElFRCwgSU5DTFVESU5HIEJVVCBOT1QgTElNSVRFRCBUTyBUSEUgV0FSUkFOVElFUyBPRiBNRVJDSEFOVEFCSUxJVFksCkZJVE5FU1MgRk9SIEEgUEFSVElDVUxBUiBQVVJQT1NFIEFORCBOT05JTkZSSU5HRU1FTlQuIElOIE5PIEVWRU5UIFNIQUxMIFRIRQpBVVRIT1JTIE9SIENPUFlSSUdIVCBIT0xERVJTIEJFIExJQUJMRSBGT1IgQU5ZIENMQUlNLCBEQU1BR0VTIE9SIE9USEVSCkxJQUJJTElUWSwgV0hFVEhFUiBJTiBBTiBBQ1RJT04gT0YgQ09OVFJBQ1QsIFRPUlQgT1IgT1RIRVJXSVNFLCBBUklTSU5HIEZST00sCk9VVCBPRiBPUiBJTiBDT05ORUNUSU9OIFdJVEggVEhFIFNPRlRXQVJFIE9SIFRIRSBVU0UgT1IgT1RIRVIgREVBTElOR1MgSU4gVEhFClNPRlRXQVJFLgoqKioqKioqKioqKioqKioqKioqKioqKioqKioqKioqKioqKioqKioqKioqKioqKioqKioqKioqKioqKioqKioqKioqKioqKioqKioqKioqLwoK0KjQu9CwINCh0LDRiNCwINC/0L4g0YjQvtGB0YHQtSDQuCDRgdC+0YHQsNC70LAg0YHRg9GI0LrRgyEK
```
Comment (sompylasar) [inline Bricks/util/golden/base64test.base64]: The `=` padding (base64) / no padding (base64url) isn't tested with this golden content.

File: Bricks/util/base64.h (hunk 3 above)
Comment (dkorolev) [inline Bricks/util/base64.h]: https://github.com/c5t/Current/pull/427

File: Bricks/util/base64.h (hunk 2 above)
Comment (dkorolev) [inline Bricks/util/base64.h]: https://github.com/c5t/Current/pull/427

File: Bricks/util/golden/base64test.base64 (hunk 4 above)
Comment (dkorolev) [inline Bricks/util/golden/base64test.base64]: Tested in the unit test though.

Comment (sompylasar) [inline Bricks/util/golden/base64test.base64]: Yep, missed that.

**Pull request merged**